consensus/state.go

```diff
@@ -262,11 +268,17 @@ func (cs *State) GetValidators() (int64, []*types.Validator) {
 	return cs.state.LastBlockHeight, cs.state.Validators.Copy().Validators
 }
 
-// SetPrivValidator sets the private validator account for signing votes.
+// SetPrivValidator sets the private validator account for signing votes. It
+// immediately requests pubkey and caches it.
 func (cs *State) SetPrivValidator(priv types.PrivValidator) {
 	cs.mtx.Lock()
+	defer cs.mtx.Unlock()
+
 	cs.privValidator = priv
-	cs.mtx.Unlock()
+
+	if err := cs.updatePrivValidatorPubKey(); err != nil {
+		cs.Logger.Error("Can't get private validator pubkey", "err", err)
+	}
```

Comment:
Should this be a hard error condition? If we can't fetch the validator's public key on startup, is there even any point in continuing with consensus?

Reply:
the very first time we try to get pubkey is in node.go https://github.com/tendermint/tendermint/blob/48426c5992c3f17c80f8c23fff84640b9f03c997/node/node.go#L664

it's fine to continue here => validator will just not sign / vote on this block (or more depending on downtime)